components/remerge/src/schema/desc.rs

```diff
@@ -69,6 +71,186 @@ pub struct Field {
     pub own_idx: FieldIndex,
 }
 
+impl Field {
+    pub fn validate(&self, v: JsonValue) -> Result<JsonValue> {
+        use InvalidRecord::*;
+        if !self.required && v.is_null() {
+            return Ok(v);
+        }
+        match &self.ty {
+            FieldType::Untyped { .. } => Ok(v),
+            FieldType::OwnGuid { .. } => {
+                if let JsonValue::String(s) = v {
+                    if s.len() < 8 || !sync_guid::Guid::from(s.as_str()).is_valid_for_sync_server()
+                    {
+                        throw!(InvalidGuid(self.name.clone()))
+                    } else {
+                        Ok(JsonValue::String(s))
+                    }
+                } else {
+                    throw!(WrongFieldType(self.name.clone(), self.ty.kind()));
+                }
+            }
+            FieldType::Text { .. } => {
+                if v.is_string() {
+                    Ok(v)
+                } else {
+                    throw!(WrongFieldType(self.name.clone(), self.ty.kind()));
+                }
+            }
+            FieldType::Boolean { .. } => {
+                if let JsonValue::Bool(b) = v {
+                    Ok(JsonValue::Bool(b))
+                } else {
+                    throw!(WrongFieldType(self.name.clone(), self.ty.kind()));
+                }
+            }
+
+            FieldType::UntypedMap { .. } => {
+                if v.is_object() {
+                    Ok(v)
+                } else {
+                    throw!(WrongFieldType(self.name.clone(), self.ty.kind()));
+                }
+            }
+
+            FieldType::RecordSet { id_key, .. } => self.validate_record_set(id_key.as_str(), v),
+            FieldType::Url { is_origin, .. } => {
+                if let JsonValue::String(s) = v {
+                    if let Ok(url) = Url::parse(&s) {
+                        if *is_origin {
+                            let o = url.origin();
+                            if !o.is_tuple() {
```

Comment:
Ah, I misunderstood the check on `is_tuple` as being the check to restrict it to the origin, rather than an additional safeguard. I would have assumed that we error out rather than truncate.

Reply:
Sorry, I see rereading the [spec](https://github.com/mozilla/application-services/pull/658/files#diff-71bb41c1e10b65f11202f8a2ea4f6278R187) that the behavior is as specified.